src/bin/orc-export.rs

```diff
@@ -13,30 +14,119 @@ struct Cli {
-    // TODO: head=N
-    // TODO: convert_dates
-    // TODO: format=[csv|json]
-    // TODO: columns="col1,col2"
+    output_file: Option<PathBuf>,
+    /// Output format. If not provided then the output is csv
+    #[arg(value_enum, short, long)]
+    format: Option<FileFormat>,
+    /// export only first N records
+    #[arg(short, long, value_name = "N")]
+    num_rows: Option<i64>,
+    /// export only provided columns. Comma separated list
+    #[arg(short, long, value_delimiter = ',')]
+    columns: Option<Vec<String>>,
+}
+
+#[derive(Clone, Debug, PartialEq, ValueEnum)]
+enum FileFormat {
+    /// Output data in csv format
+    Csv,
+    /// Output data in json format
+    Json,
+}
+
+enum OutputWriter<W: io::Write, F: JsonFormat> {
+    Csv(csv::Writer<W>),
+    Json(json::Writer<W, F>),
+}
+
+impl<W, F> OutputWriter<W, F>
+where
+    W: io::Write,
+    F: JsonFormat,
+{
+    fn write(&mut self, batch: &RecordBatch) -> Result<(), ArrowError> {
+        match self {
+            OutputWriter::Csv(w) => w.write(batch),
+            OutputWriter::Json(w) => w.write(batch),
+        }
+    }
+
+    fn finish(&mut self) -> Result<(), ArrowError> {
+        match self {
+            OutputWriter::Csv(_) => Ok(()),
+            OutputWriter::Json(w) => w.finish(),
+        }
+    }
 }
 
 fn main() -> Result<()> {
     let cli = Cli::parse();
-    let f = File::open(&cli.file)?;
-    let output_writer: Box<dyn io::Write> = if let Some(output) = cli.output {
+
+    // Prepare reader
+    let mut f = File::open(&cli.file)?;
+    let metadata = read_metadata(&mut f)?;
+    let reader = ArrowReaderBuilder::try_new(f)?.build();
+
+    // Prepare writer
+    let writer: Box<dyn io::Write> = if let Some(output) = cli.output_file {
         Box::new(File::create(output)?)
     } else {
         Box::new(io::stdout())
     };
 
-    let reader = ArrowReaderBuilder::try_new(f)?.build();
-    let mut writer = csv::WriterBuilder::new()
-        .with_header(true)
-        .build(output_writer);
+    let mut output_writer = match cli.format {
+        Some(FileFormat::Json) => {
+            OutputWriter::Json(json::WriterBuilder::new().build::<_, LineDelimited>(writer))
+        }
+        _ => OutputWriter::Csv(csv::WriterBuilder::new().with_header(true).build(writer)),
+    };
+
+    // Select columns which should be exported (Binary and Decimal are not supported)
+    let is_json = cli.format == Some(FileFormat::Json);
+    let skip_cols: Vec<usize> = metadata
+        .root_data_type()
+        .children()
+        .iter()
+        .enumerate()
+        .filter(|(_, nc)| match nc.data_type().to_arrow_data_type() {
+            DataType::Binary => true,
+            DataType::Decimal128(_, _) | DataType::Decimal256(_, _) => is_json,
+            _ => {
+                if let Some(cols) = &cli.columns {
+                    !cols.iter().any(|c| nc.name().eq(c))
+                } else {
+                    false
+                }
```

Comment:
Can try push this to use the projection logic provided by ArrowReaderBuilder

Reply:
Done